File: provider/src/rpc.rs
use jsonrpsee::proc_macros::rpc;
use std::collections::HashMap;

use super::response::*;
use crate::web_socket::ws_message::message::request::ClientOffset;
use crate::web_socket::ws_message::message::response::TxTopicEvent;
use crate::web_socket::ws_message::topic::Topic;
use jsonrpsee::core::RpcResult;
use zklink_sdk_types::basic_types::tx_hash::TxHash;
use zklink_sdk_types::basic_types::{AccountId, BlockNumber, ChainId, SubAccountId, TokenId, ZkLinkAddress};
use zklink_sdk_types::prelude::BigUintSerdeWrapper;
use zklink_sdk_types::signatures::TxLayer1Signature;
use zklink_sdk_types::tx_type::zklink_tx::{ZkLinkTx, ZkLinkTxType};

#[rpc(client, server)]
pub trait ZkLinkRpc {
    #[method(name = "getSupportChains")]
    async fn get_support_chains(&self) -> RpcResult<Vec<ChainResp>>;

    #[method(name = "getSupportTokens")]
    async fn tokens(&self) -> RpcResult<HashMap<TokenId, TokenResp>>;

    #[method(name = "getLatestBlockNumber")]
    async fn block_info(&self) -> RpcResult<BlockNumberResp>;

    #[method(name = "getBlockByNumber")]
    async fn block_detail(
        &self,
        block_number: Option<BlockNumber>,
        include_tx: bool,
        include_update: bool,
    ) -> RpcResult<BlockResp>;

    #[method(name = "getPendingBlock")]
    async fn pending_block_detail(
        &self,
        last_tx_timestamp_micro: u64,
        include_tx: bool,
        include_update: bool,
        limit: Option<usize>,
    ) -> RpcResult<Vec<TxHashOrDetailResp>>;

    #[method(name = "getBlockOnChainByNumber")]
    async fn block_onchain_detail(&self, block_number: BlockNumber) -> RpcResult<BlockOnChainResp>;

    #[method(name = "getAccount")]
    async fn account_info(&self, account_query: AccountQuery) -> RpcResult<AccountInfoResp>;

    #[method(name = "getSubAccountGlobalVars")]
    async fn global_vars_info(&self, sub_account_query: SubAccountId) -> RpcResult<GlobalVarsResp>;

    #[method(name = "getAccountBalances")]
    async fn account_balances(
        &self,
        account_id: AccountId,
        sub_account_id: Option<SubAccountId>,
    ) -> RpcResult<SubAccountBalances>;

    #[method(name = "getAccountOrderSlots")]
    async fn account_order_slots(
        &self,
        account_id: AccountId,
        sub_account_id: Option<SubAccountId>,
    ) -> RpcResult<SubAccountOrders>;

    #[method(name = "getAccountPositions")]
    async fn account_positions(
        &self,
        account_id: AccountId,
        sub_account_id: Option<SubAccountId>,
    ) -> RpcResult<SubAccountPositions>;

    #[method(name = "getTokenReserve")]
    async fn token_remain(&self, token_id: TokenId, mapping: bool) -> RpcResult<HashMap<ChainId, BigUintSerdeWrapper>>;

    #[method(name = "getAccountSnapshot")]
    async fn account_snapshot(
        &self,
        account_query: AccountQuery,
        sub_account_id: Option<SubAccountId>,
        block_number: Option<BlockNumber>,
    ) -> RpcResult<AccountSnapshotResp>;

    #[method(name = "getTransactionByHash")]
    async fn tx_info(&self, hash: TxHash, include_update: bool) -> RpcResult<TxResp>;

    #[method(name = "getAccountTransactionHistory")]
    async fn tx_history(
        &self,
        tx_type: ZkLinkTxType,
        address: ZkLinkAddress,
        page_index: u64,
        page_size: u32,
    ) -> RpcResult<Page<ZkLinkTxHistory>>;

    #[method(name = "getWithdrawTxs")]
    async fn tx_withdraw(&self, last_tx_timestamp_micro: u64, max_txs: u32) -> RpcResult<Vec<WithdrawTxResp>>;

    #[method(name = "getWebSocketEvents")]
    async fn get_websocket_events(&self, topic: Topic, offset: ClientOffset) -> RpcResult<Vec<TxTopicEvent>>;

    #[method(name = "getChangePubkeyChainId")]
    async fn get_change_pubkey_chain_id(&self) -> RpcResult<ChainId>;

    #[method(name = "getEthProperty")]
    async fn get_eth_property(&self) -> RpcResult<EthPropertyResp>;

    #[method(name = "sendTransaction")]
    async fn tx_submit(&self, tx: ZkLinkTx, l1_signature: Option<TxLayer1Signature>) -> RpcResult<TxHash>;
}

File: examples/Javascript/js-example/rpc.js
import init, *  as wasm  from "./web-dist/zklink-sdk-web.js";
async function main() {
    await init();
    try {
        let provider = new wasm.Provider("mainet");
        // 1.getSupportTokens
        let tokens = await provider.getSupportTokens();
        console.log(tokens);
        // 2.getAccountSnapshot
        let account_id = new wasm.AccountQueryParam(wasm.AccountQueryType.AccountId, "5");
        let sub_account_id = 1;
        // let block_number = 100;
        let account_resp = await provider.getAccountSnapshot(account_id,sub_account_id,null);
        console.log(account_resp);
        // 3.sendTransaction(test on the tx example)
        // 4.getSupportChains
        let chains = await provider.getSupportChains();
        console.log(chains);
        // 5.getLatestBlockNumber
        let block_info = await provider.getLatestBlockNumber();
        console.log(block_info);
        // 6.getBlockByNumber
        let block_detail = await provider.getBlockByNumber(100,true,true);
        console.log(block_detail);
        // 7.getPendingBlock
        let pending_block_info = await provider.getPendingBlock(1696743981000n,true,true,null);
        console.log(pending_block_info);
        // 8.getBlockOnChainByNumber
        let on_chain_block_info = await provider.getBlockOnChainByNumber(100);
        console.log(on_chain_block_info);
        // 9.getAccount
        let get_account_id = new wasm.AccountQueryParam(wasm.AccountQueryType.AccountId, "10");
        let account = await provider.getAccount(get_account_id);
        console.log(account);
        // 10.getAccountBalances
        let balances = await provider.getAccountBalances(20,1);
        console.log(balances);
        // 11.getAccountOrderSlots
        let slots = await provider.getAccountOrderSlots(20,1);
        console.log(slots);
        // 12.getTokenReserve
        let reserve = await provider.getTokenReserve(18,false);
        console.log(reserve);
        // 13.getTransactionByHash
        let tx_hash = "0x0cbeabac1a2257fb095c2465e148570e32793345442b39bf64cad4ed87475f9b";
        let tx_info = await provider.getTransactionByHash(tx_hash,false);
        console.log(tx_info);
        // 14.getAccountTransactionHistory
        let history = await provider.getAccountTransactionHistory(wasm.L2TxType.Deposit,"0x12aFF993702B5d623977A9044686Fa1A2B0c2147",0n,5);
        console.log(history);
        // 15.getFastWithdrawTxs
        let fast_withdraw_txs = await provider.getFastWithdrawTxs(1696743981000n,10);
        console.log(fast_withdraw_txs);
        // 16.pullForwardTxs
        // 17.confirmFullExit
    } catch (error) {
        console.error(error);
    }

}

main();
